src/actix/api_key.rs

```diff
@@ -71,3 +93,28 @@ where
         })
     }
 }
+
+fn is_read_only(req: &ServiceRequest) -> bool {
+    static READ_ONLY_PATHS: OnceLock<Vec<Regex>> = OnceLock::new();
+
+    let read_only_paths = READ_ONLY_PATHS.get_or_init(|| {
+        vec![
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/count/?(\?.*)?$").expect("bad regex"),
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/search/?(\?.*)?$").expect("bad regex"),
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/search/batch/?(\?.*)?$")
+                .expect("bad regex"),
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/recommend/?(\?.*)?$")
+                .expect("bad regex"),
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/recommend/groups/?(\?.*)?$")
+                .expect("bad regex"),
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/recommend/batch/?(\?.*)?$")
+                .expect("bad regex"),
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/?(\?.*)?$").expect("bad regex"),
+            Regex::new(r"^/collections/[a-z0-9-_&;]+/points/scroll/?(\?.*)?$").expect("bad regex"),
+        ]
+    });
+
+    req.method() == Method::GET
+        || (req.method() == Method::POST
+            && read_only_paths.iter().any(|re| re.is_match(req.path())))
```

Comment:
This can be parallelized with rayon. It might be worth benchmarking the parallel version if this ends up being the chosen approach.